### PR TITLE
[dev] Support request headers override in middleware

### PR DESCRIPTION
**Pull request by nuta**
Implements request headers override in middlewares.

#### New middleware headers

- `x-middleware-override-headers`: A comma separated list of *all* request header names. Headers not listed will be deleted.
- `x-middleware-request-<name>`: A new value for the header `<name>`.

### Related Issues

- #8724: Add helper functions for non-Next.js middlewares
- https://github.com/vercel/next.js/pull/41380: Next.js' implementation

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/cli/src/util/dev/server.ts [hunk 1]
```diff
@@ -1478,6 +1478,29 @@ export default class DevServer {
...
+              if (newValue) {
+                req.headers[name] = newValue;
+              } else {
+                delete req.headers[name];
+              }
+
+              middlewareRes.headers.delete(valueKey);
+            }
+            middlewareRes.headers.delete('x-middleware-override-headers');
+          }
```
Comment (Schniz) [inline packages/cli/src/util/dev/server.ts]: do we want to extract this into a simple to understand function, that can easily be unit testable? `applyOverriddenHeaders(middlewareRes.headers)` <picture data-single-emoji=":eyeses:" title=":eyeses:">:eyeses:</picture>

Comment (EndangeredMassa) [reply packages/cli/src/util/dev/server.ts]: I agree. Some unit tests for the new logic would be great in this or a follow-up PR.

**Review by EndangeredMassa — APPROVED** [review]

**Review by javivelasco — APPROVED** [review]

File: packages/cli/src/util/dev/headers.ts [hunk 2]
```diff
@@ -16,3 +16,60 @@ export function nodeHeadersToFetchHeaders(
...
+  respHeaders: Headers,
+  skippedHeaders: Set<string>
+) {
+  if (!respHeaders.has('x-middleware-override-headers')) {
+    return;
+  }
+
+  const overriddenHeaders: Set<string> = new Set();
+  for (const key of respHeaders
+    .get('x-middleware-override-headers')!
```
Comment (Ethan-Arrowood) [inline packages/cli/src/util/dev/headers.ts]: Is there an edge case where `respHeaders.get('x-middleware-override-headers')` could return `null` or `undefined` or some that isn't a string?

**Review by Ethan-Arrowood — APPROVED** [review]
lgtm - see potential edge case comment. might just need a simple guard.

File: packages/cli/src/util/dev/headers.ts (hunk 2 above)
Comment (nuta) [reply packages/cli/src/util/dev/headers.ts]: No there was a if guard to ensure it exists. Refactored that part so that we don't need to use `!` 👍

**Pull request merged**